### ISSUE TITLE
Type incorrectly inferred under very specific circumstances involving a type alias for a callable.

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#1968]
Pyright incorrectly infers the type under some specific circumstances involving:
-  a type alias for a callable that returns a union of objects that implement the same generic type, ie. `alias = Callable[[], Union[T, Wrapper[T]]`. 
- a function that takes a function as argument, typed with above alias, setting the generic type using a `TypeVar`, and returns an object that also includes that `TypeVar` as the generic type: ie. `def some_func(func: alias[T]) -> Wrapper[T]:`
- a second function that takes an argument typed with the same union in the above alias example, that returns the wrapping object. ie. `def other_func(item: Union[T, Wrapper[T]]) -> Wrapper[T]`:

Full example that showcases the bug:
```py
from typing import Callable, Generic, TypeVar, Union

T = TypeVar("T")


class Wrapper(Generic[T]):
    def __init__(self, inner: T):
        self.inner = inner


FuncAlias = Callable[[], Union[T, Wrapper[T]]]


def some_func(func: FuncAlias[T]) -> Wrapper[T]:
    return ensure_wrapped(func())
    # Expression of type "Wrapper[Wrapper[T@some_func]]" cannot be assigned to return type "Wrapper[T@some_func]"


def ensure_wrapped(item: Union[T, Wrapper[T]]) -> Wrapper[T]:
    if isinstance(item, Wrapper):
        return item
    return Wrapper(item)


```

This bug must have been recently introduced in pyright. For reference I know version `1.1.143` correctly infers the type:

```py
def some_func(func: FuncAlias[T]) -> Wrapper[T]:
    reveal_type(ensure_wrapped(func())) # Type of "ensure_wrapped(func())" is "Wrapper[T@some_func]"
    return ensure_wrapped(func())
```

Bug was first observed on pyright version `1.1.147`, but still present as of `1.1.148`. (Jumped from `1.1.143` to `1.1.147` so may have been introduced in some version in between).

**Comment by kennipj** [comment]
On further inspection, I just realised this probably isn't a bug, but an oversight on my part. (If `T` is of type `Wrapper`, then `some_func` would return `Wrapper[T]`, but according to type info, it should return `Wrapper[Wrapper[T]]`).

**Comment by erictraut** [comment]
Thanks for reporting. This is an interesting case because there are two valid solutions for `T` when considering the type expression `Union[T, Wrapper[T]]` and an argument of type `Wrapper[X]`. The two valid solutions are `T = Wrapper[X]` and `T = X`. The problem here is that it is choosing `T = Wrapper[X]`, and you're expecting it to choose the other. I have some heuristics in place to try to pick the simplest answer. I must have broken one of these heuristics with another recent change. I'll take a look at it.

**Comment by erictraut** [comment]
As anticipated, a recent change exposed a weakness in the heuristics that I have in place to handle this ambiguous case. The heuristic measures the "complexity" of each possible solution for `T` and picks the least complex result. I've updated the code that computes this "complexity score" so it picks the desired solution in this case. I've also added a new unit test based on your code so we can avoid future regressions of this sort.
